libraries/AP_Proximity/AP_Proximity_Cygbot_D1.cpp

```diff
@@ -0,0 +1,188 @@
+		// try sending the sensor the "send data" message
+		_initialized = false;
+		set_status(AP_Proximity::Status::NoData);
+	}
+}
+
+// send message to the sensor to start streaming 2-D data
+void AP_Proximity_Cygbot_D1::send_sensor_start()
+{
+	// this message corresponds to "start message"
+	uint8_t PACKET_START_2D[8] = { CYGBOT_PACKET_HEADER_0, CYGBOT_PACKET_HEADER_1, CYGBOT_PACKET_HEADER_2, 0x02, 0x00, 0x01, 0x00, 0x03 };
+	_uart->write(PACKET_START_2D, 8);
+}
+
+void AP_Proximity_Cygbot_D1::read_sensor_data()
+{
+	uint32_t nbytes = _uart->available();
+
+    while (nbytes-- > 0) {
+        uint8_t byte = _uart->read();
+		if (!parse_byte(byte)) {
+			// reset
+			reset();
+		}
+	}
+}
+
+// parse one byte from the senor. Return false on error.
+// Message format is: header1 + header2 + header3 + length1 + length2 + PayloadCommand + checksum
+bool AP_Proximity_Cygbot_D1::parse_byte(uint8_t data)
+{
+	switch (_parse_state)
+	{
+	case Header1:
+		if (data == CYGBOT_PACKET_HEADER_0) {
+			_parse_state = Header2;
+			return true;
+		}
+
+		return false;
+
+	case Header2:
+		if (data == CYGBOT_PACKET_HEADER_1) {
+			_parse_state = Header3;
+			return true;
+		}
+		return false;
+
+	case Header3:
+		if (data == CYGBOT_PACKET_HEADER_2) {
+			_parse_state = Length1;
+			return true;
+		}
+		return false;
+
+	case Length1:
+		_msg.payload_len_flags_low = data;
+		_parse_state = Length2;
+		return true;
+
+	case Length2:
+		_msg.payload_len_flags_high = data;
+		_msg.payload_len = UINT16_VALUE(data, _msg.payload_len_flags_low);
+		if (_msg.payload_len > CYGBOT_MAX_MSG_SIZE) {
+			return false;
+		}
+		_parse_state = Payload_Header;
+		return true;
+
+	case Payload_Header:
+		if (data == CYGBOT_PAYLOAD_HEADER) {
+			_parse_state = Payload_Data;
+			_msg.payload_counter = 0;
+			_msg.payload[_msg.payload_counter] = data;
+			_msg.payload_counter ++;
+			return true;
+		}
+		return false;
+
+	case Payload_Data:
+		if (_msg.payload_counter < (_msg.payload_len - 1)) {
+			_msg.payload[_msg.payload_counter] = data;
+			_msg.payload_counter++;
+			return true;
+		}
+		_parse_state = CheckSum;
+		return true;
+
+	case CheckSum:
+		{
+			// To-DO: FIX CHECKSUM below. It does not pass correctly
+			// const uint8_t checksum_num = calc_checksum(_msg.payload, _msg.payload_len);
+			// if (data != checksum_num) {
+			// 	return false;
+			// }
+			// checksum is valid, parse payload
+			_last_distance_received_ms = AP_HAL::millis();
+			parse_payload();
+			_temp_boundary.update_3D_boundary(boundary);
+			reset();
+			return true;
+		}
+		break;
+
+	default:
+		return false;
+	}
+
+	return false;
+}
+
+// parse payload, to pick out distances, and feed them to the correct faces
+void AP_Proximity_Cygbot_D1::parse_payload()
+{
+	// current horizontal angle in the payload
+	float sampled_angle = CYGBOT_2D_START_ANGLE;
+
+	for (uint16_t i = 1; i < _msg.payload_len - 1; i += 2) {
+		uint16_t distance = UINT16_VALUE(_msg.payload[i], _msg.payload[i+1]);
```

Comment:
I guess this could be made "const" and change the local variable from "distance" to "distance_mm".

All the indenting seems to be 2 spaces instead of our regular 4.